File: lib/utils/errorHandler.ts
import { logger } from './logger';

export enum ErrorType {
  NETWORK = 'network',
  PERMISSION = 'permission',
  CONNECTION = 'connection',
  VALIDATION = 'validation',
  MEDIA_DEVICE = 'media_device',
  ENCRYPTION = 'encryption',
  UNKNOWN = 'unknown'
}

export interface ErrorInfo {
  type: ErrorType;
  message: string;
  originalError: unknown;
  userMessage: string;
  canRetry: boolean;
}

/**
 * Classifies an error and returns structured error information
 */
function classifyError(error: unknown): ErrorInfo {
  const errorMessage = error instanceof Error ? error.message : String(error);
  const errorMessageLower = errorMessage.toLowerCase();
  
  // Network errors
  if (errorMessageLower.includes('network') || 
      errorMessageLower.includes('fetch') ||
      errorMessageLower.includes('timeout') ||
      errorMessageLower.includes('connection') && errorMessageLower.includes('failed')) {
    return {
      type: ErrorType.NETWORK,
      message: errorMessage,
      originalError: error,
      userMessage: 'Network connection issue. Please check your internet connection and try again.',
      canRetry: true
    };
  }
  
  // Permission errors
  if (errorMessageLower.includes('permission') || 
      errorMessageLower.includes('notallowed') ||
      errorMessageLower.includes('denied')) {
    return {
      type: ErrorType.PERMISSION,
      message: errorMessage,
      originalError: error,
      userMessage: 'Permission denied. Please allow access to your camera and microphone.',
      canRetry: true
    };
  }
  
  // Connection errors
  if (errorMessageLower.includes('webrtc') ||
      errorMessageLower.includes('ice') ||
      errorMessageLower.includes('rtcpeerconnection')) {
    return {
      type: ErrorType.CONNECTION,
      message: errorMessage,
      originalError: error,
      userMessage: 'Connection error. Please refresh the page and try again.',
      canRetry: true
    };
  }
  
  // Media device errors
  if (errorMessageLower.includes('camera') ||
      errorMessageLower.includes('microphone') ||
      errorMessageLower.includes('mediastream') ||
      errorMessageLower.includes('getusermedia')) {
    return {
      type: ErrorType.MEDIA_DEVICE,
      message: errorMessage,
      originalError: error,
      userMessage: 'Media device error. Please check your camera and microphone settings.',
      canRetry: true
    };
  }
  
  // Encryption errors
  if (errorMessageLower.includes('encryption') ||
      errorMessageLower.includes('e2ee')) {
    return {
      type: ErrorType.ENCRYPTION,
      message: errorMessage,
      originalError: error,
      userMessage: 'Encryption error. Your browser may not support encrypted meetings.',
      canRetry: false
    };
  }
  
  // Validation errors
  if (errorMessageLower.includes('validation') ||
      errorMessageLower.includes('invalid')) {
    return {
      type: ErrorType.VALIDATION,
      message: errorMessage,
      originalError: error,
      userMessage: 'Invalid request. Please check your input and try again.',
      canRetry: false
    };
  }
  
  // Unknown errors
  return {
    type: ErrorType.UNKNOWN,
    message: errorMessage,
    originalError: error,
    userMessage: 'An unexpected error occurred. Please try again later.',
    canRetry: true
  };
}

/**
 * Main error handler function
 * @param error The error to handle
 * @param context Context information about where the error occurred
 * @returns Structured error information
 */
export function handleError(error: unknown, context: string): ErrorInfo {
  const errorInfo = classifyError(error);
  
  // Log the error with context
  logger.error(`Error in ${context}:`, {
    type: errorInfo.type,
    message: errorInfo.message,
    canRetry: errorInfo.canRetry
  });
  
  return errorInfo;
}

/**
 * Handle API errors specifically
 */
export async function handleApiError(response: Response, context: string): Promise<ErrorInfo> {
  let errorMessage = `API error: ${response.status} ${response.statusText}`;
  
  try {
    const data = await response.json();
    errorMessage = data.message || data.error || errorMessage;
  } catch {
    // Response body is not JSON or already consumed
  }
  
  const error = new Error(errorMessage);
  return handleError(error, context);
}

/**
 * Create a user-friendly error message
 */
export function getUserFriendlyErrorMessage(error: unknown): string {
  const errorInfo = classifyError(error);
  return errorInfo.userMessage;
}

/**
 * Check if an error is retryable
 */
export function isRetryableError(error: unknown): boolean {
  const errorInfo = classifyError(error);
  return errorInfo.canRetry;
}

File: almajd_app/lib/utils/error_handler.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import '../utils/logger.dart';

/// Error categories for better error handling
enum ErrorCategory {
  network,
  permission,
  authentication,
  authorization,
  server,
  timeout,
  validation,
  unknown,
}

/// Global error handler for the app
/// Provides user-friendly error messages and error reporting
class ErrorHandler {
  /// Handle and display user-friendly error messages
  static void handleError(
    BuildContext? context,
    dynamic error, {
    StackTrace? stackTrace,
    String? userMessage,
    bool showSnackBar = true,
    bool reportToCrashlytics = true,
    VoidCallback? onRetry,
  }) {
    // Log the error
    Logger.error(
      userMessage ?? 'An error occurred: $error',
      error,
      stackTrace,
      'ErrorHandler',
    );

    // Categorize error
    final category = _categorizeError(error);
    final message = _getUserFriendlyMessage(error, category);
    final canRetry = _canRetry(category);

    // Show user-friendly message if context is available
    if (context != null && showSnackBar && context.mounted) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Row(
            children: [
              Icon(_getErrorIcon(category), color: Colors.white),
              const SizedBox(width: 12),
              Expanded(child: Text(message)),
            ],
          ),
          backgroundColor: _getErrorColor(category),
          behavior: SnackBarBehavior.floating,
          duration: const Duration(seconds: 4),
          action: (canRetry && onRetry != null)
              ? SnackBarAction(
                  label: 'Retry',
                  textColor: Colors.white,
                  onPressed: onRetry,
                )
              : null,
        ),
      );
    }

    // In production, send to error reporting service
    if (kReleaseMode && reportToCrashlytics) {
      // TODO: Integrate with Firebase Crashlytics or Sentry
      // _reportToCrashlytics(error, stackTrace);
    }
  }

  /// Categorize error for better handling
  static ErrorCategory _categorizeError(dynamic error) {
    final errorString = error.toString().toLowerCase();

    if (errorString.contains('permission')) {
      return ErrorCategory.permission;
    } else if (errorString.contains('network') || 
               errorString.contains('connection') ||
               errorString.contains('socket')) {
      return ErrorCategory.network;
    } else if (errorString.contains('timeout')) {
      return ErrorCategory.timeout;
    } else if (errorString.contains('unauthorized') || 
               errorString.contains('401') ||
               errorString.contains('authentication')) {
      return ErrorCategory.authentication;
    } else if (errorString.contains('forbidden') || 
               errorString.contains('403') ||
               errorString.contains('authorization')) {
      return ErrorCategory.authorization;
    } else if (errorString.contains('server') || 
               errorString.contains('500') ||
               errorString.contains('502') ||
               errorString.contains('503')) {
      return ErrorCategory.server;
    } else if (errorString.contains('validation') ||
               errorString.contains('invalid') ||
               errorString.contains('not found')) {
      return ErrorCategory.validation;
    } else {
      return ErrorCategory.unknown;
    }
  }

  /// Get user-friendly error message from error object
  static String _getUserFriendlyMessage(dynamic error, ErrorCategory category) {
    switch (category) {
      case ErrorCategory.network:
        return 'Network error. Please check your internet connection and try again.';
      case ErrorCategory.permission:
        return 'Permission denied. Please enable required permissions in settings.';
      case ErrorCategory.timeout:
        return 'Request timed out. Please check your connection and try again.';
      case ErrorCategory.authentication:
        return 'Authentication failed. Please log in again.';
      case ErrorCategory.authorization:
        return 'Access denied. You don\'t have permission for this action.';
      case ErrorCategory.server:
        return 'Server error. Please try again in a moment.';
      case ErrorCategory.validation:
        final errorString = error.toString().toLowerCase();
        if (errorString.contains('room') && errorString.contains('not found')) {
          return 'Room not found. Please check the room name.';
        }
        return 'Invalid input. Please check your information and try again.';
      case ErrorCategory.unknown:
        return 'An unexpected error occurred. Please try again.';
    }
  }

  /// Check if error can be retried
  static bool _canRetry(ErrorCategory category) {
    return category == ErrorCategory.network ||
           category == ErrorCategory.timeout ||
           category == ErrorCategory.server;
  }

  /// Get error icon based on category
  static IconData _getErrorIcon(ErrorCategory category) {
    switch (category) {
      case ErrorCategory.network:
        return Icons.wifi_off;
      case ErrorCategory.permission:
        return Icons.lock_outline;
      case ErrorCategory.timeout:
        return Icons.schedule;
      case ErrorCategory.authentication:
        return Icons.login;
      case ErrorCategory.authorization:
        return Icons.block;
      case ErrorCategory.server:
        return Icons.cloud_off;
      case ErrorCategory.validation:
        return Icons.error_outline;
      case ErrorCategory.unknown:
        return Icons.error_outline;
    }
  }

  /// Get error color based on category
  static Color _getErrorColor(ErrorCategory category) {
    switch (category) {
      case ErrorCategory.network:
      case ErrorCategory.timeout:
        return Colors.orange;
      case ErrorCategory.permission:
      case ErrorCategory.authorization:
        return Colors.amber;
      case ErrorCategory.authentication:
        return Colors.blue;
      case ErrorCategory.server:
      case ErrorCategory.validation:
      case ErrorCategory.unknown:
        return Colors.red;
    }
  }

  /// Handle network errors specifically
  static void handleNetworkError(BuildContext? context, dynamic error) {
    handleError(
      context,
      error,
      userMessage: 'Network error. Please check your internet connection.',
    );
  }

  /// Handle permission errors specifically
  static void handlePermissionError(BuildContext? context, String permission) {
    handleError(
      context,
      'Permission denied: $permission',
      userMessage: 'Permission denied. Please enable $permission in settings.',
    );
  }

  /// Handle room validation errors
  static void handleRoomError(BuildContext? context, dynamic error) {
    handleError(
      context,
      error,
      userMessage: 'Room error. Please check the room name and try again.',
    );
  }
}



